cmd/gazelle/gazelle.go

```diff
@@ -59,6 +62,29 @@ func main() {
 	log.SetPrefix("gazelle: ")
 	log.SetFlags(0) // don't print timestamps
 
+	if cpuProfile := os.Getenv("GAZELLE_CPU_PROFILE"); cpuProfile != "" {
```

Comment:
I took a quick crack at it and found it to be a bit difficult to move to flag.

The reason is most of the flags are handled by the subcommands' implementations of `config.Configurer.RegisterFlags()` interface while I wanted to have the profiling functionality in the top level of the gazelle binary, preferably as part of `main()`.

The initial attempt resulted in various conflicts between different flag sets.
I thought of making profiling flags exclusive to args before the subcommand, so `gazelle [--profile_path=] fix [--fix-flags-here=]`... but turn out we put some subcommand flags before the subcommand itself (i.e. in `go_repository` https://github.com/bazelbuild/bazel-gazelle/blob/1ca72fdef279b2a7b0fbef22572aece1380361a9/internal/go_repository.bzl#L259)

I will give this another try when I have time. Helps / pointers would be much appreciated.

Reply:
I think you could create a [new `flag.FlagSet`](https://pkg.go.dev/flag?utm_source=godoc#NewFlagSet), set the error mode to [`ContinueOnError`](https://pkg.go.dev/flag?utm_source=godoc#ErrorHandling) define a [`Bool` flag](https://pkg.go.dev/flag?utm_source=godoc#FlagSet.Bool) and [`flagSet.Parse`](https://pkg.go.dev/flag?utm_source=godoc#FlagSet.Parse) the flags. Then you can discard the error (since not all flags will be defined) and use the resulting `bool` to make a determination on what to do.

Alternately, and this might really be the way to go, you could make a helper function for doing profiling in the way that is expected and explicitly add the profiler to the flags for each sub command. This is likely the direction I would go.

Reply:
```
--- a/cmd/gazelle/gazelle.go
+++ b/cmd/gazelle/gazelle.go
@@ -22,6 +22,8 @@ import (
        "fmt"
        "log"
        "os"
+       "runtime"
+       "runtime/pprof"

        "github.com/bazelbuild/bazel-gazelle/config"
        "github.com/bazelbuild/bazel-gazelle/language"
@@ -59,6 +61,31 @@ func main() {
        log.SetPrefix("gazelle: ")
        log.SetFlags(0) // don't print timestamps

+       var (
+               cpuProfilePath string
+               cpuProfileRate int
+       )
+       profileFs := flag.NewFlagSet("profiling", flag.ContinueOnError)
+       profileFs.Usage = func() {}
+       profileFs.StringVar(&cpuProfilePath, "cpu_profile", "", "path to store CPU profile")
+       profileFs.IntVar(&cpuProfileRate, "cpu_profile_rate", 4096, "rate of sampling when profile CPU")
+       err := profileFs.Parse(os.Args[1:])
+       if err != nil {
+               log.Fatal("could not parse", err)
+       }
+
+       if cpuProfilePath != "" {
+               runtime.SetCPUProfileRate(cpuProfileRate)
+               f, err := os.Create(cpuProfilePath)
+               if err != nil {
+                       log.Fatal("could not create pprof output", err)
+               }
+
+               pprof.StartCPUProfile(f)
+               defer pprof.StopCPUProfile()
+       }
+       log.Printf("DEBUG: %s", profileFs.Args())
+
        var wd string
        if wsDir := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); wsDir != "" {
                wd = wsDir
@@ -69,7 +96,7 @@ func main() {
                }
        }

-       if err := run(wd, os.Args[1:]); err != nil && err != flag.ErrHelp {
+       if err := run(wd, profileFs.Args()); err != nil && err != flag.ErrHelp {
                if err == errExit {
                        os.Exit(1)
                } else {
```

This is what I have this morning, which seems to be on the right path with what you are suggesting.
Let me take a closer look at this when I have a bit more time 👀

Reply:
I really do think this would be a simple add to the individual commands. It would probably be quite a bit less painful

Reply:
I pushed a version of my patch to this PR.

It has bugs:

```
# This does work
bazel run //cmd/gazelle -- -mode=diff --cpu_profile_path=/tmp/gazelle.pprof 

# This does NOT work
bazel run //cmd/gazelle -- --cpu_profile_path=/tmp/gazelle.pprof -mode=diff
bazel run //cmd/gazelle -- -mode diff --cpu_profile_path=/tmp/gazelle.pprof 
```

This is because I blindly assume when my flag set fails to parse, the bad flag I need to remove is `args[0]`, which is not always the case. The package `flag` in Go is rather limited in how we could parse things, but yanking that out now to replace it with spf13/viper is out-of-scope.

The best approach I could think of here is to write my own flag parser. Which seems to be the wrong rabbit hole to head down.

---

I considered adding this to individual commands but I think profiling should be done from the top level to give you a good top-down picture. Having the flag in a subcommand kinda defeats the purpose.

With all that said, @achew22 would you consider the first environment variable approach as this is only meant for development/troubleshooting?